data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,452 @@
+(executable
+ (name main)
+ (preprocess
+  (action
+   (system "./preprocessor.sh %{input-file}"))))
+```
+
+### The limits of manipulating text file
+
+The complexities of a programming language syntax makes it very hard to
+manipulate text in a way that is tied to the syntax. Suppose for instance that,
+similarly to the previous example, you want to rewrite all occurences of "World"
+by "Universe", but _inside the OCaml strings of the program_ only. It is quite
+involved and requires a good knowledge of the OCaml syntax to do so, as there
+are several delimiters for strings (such as `{| ...|}`), and linebreaks or
+comments could get in the way...
+
+Consider another example. Suppose you have defined a type, and you want to
+generate at compile-time a serializer from this specific type, to an encoding of
+it in a `json` format, such as `Yojson` (see
+[here](#why-are-ppxes-especially-useful-in-ocaml) why it has to be generated at
+compile-time). This serialization code could be written by a preprocessor, which
+would look for the type in the file, and serialize it differently depending on
+the structure of the type, that is whether it is a variants type, a record type,
+the structure of its subtypes...
+
+But, to understand the structure of the type, the preprocessor need to parse
+those information. Given that parsing OCaml syntax is complex, it is not
+something that we want to do do again and again whenever we write a new
+preprocessor.
+
+The solution to this is to use PPXes, preprocessors that run on the result of
+the parsing of the file.
+
+## PPXes
+
+PPxes are a different kind of preprocessors, that do not run on the textual
+source code, but on the result of the parsing: the Abstract Syntax Tree (AST),
+or more precisely in OCaml, the parsetree. In order to understand PPXes well, we
+need to understand what is this parsetree.
+
+### OCaml's parsetree, the first OCaml's AST
+
+During the compilation phase, OCaml's compiler will parse the input file into an
+internal representation of it, called the parsetree. The program is represented
+as a tree, with a very complicated OCaml type, that you can find in the
+[`Parsetree` module](https://v2.ocaml.org/api/compilerlibref/Parsetree.html).
+
+Let us give already a few properties of this tree:
+- Each node in the AST has a type corresponding to a different role, such as
+  "let definitions", "expressions", "patterns", ...
+- The root of the tree is a list of `structure_item`s
+- A `structure_item` can either represent a toplevel expression, a type
+  definition, a let definition, ... This is determined using a variant type.
+
+There are two complementary ways of getting a grasp on the parsetree type. One
+is to read the API documentation, which include examples of what each type and
+value represent. The other is to make OCaml dump the parsetree value of crafted
+OCaml code. This is possible thanks to the option `-dparsetree`, available in
+the `ocaml` toplevel (as well as in `utop`):
```

Comment:
```suggestion
the OCaml toplevel (as well as in UTop):
```

Reply:
Here I really mean the toplevel run byt the `ocaml` command...